File: nodejs/apis/activity_enroll.js
'use strict';
/** 活动报名 API（6 端点，来自 05_activity_enroll.json） */
const { BaseRequest } = require('../base_request');
const BASE = 'https://mms.pinduoduo.com';
const ACT_REFERER = 'https://mms.pinduoduo.com/activity';

class ActivityEnrollAPI extends BaseRequest {
    async checkLogin() {
        return this.post(`${BASE}/janus/api/checkLogin`, { data: '', referer: ACT_REFERER });
    }
    async getActivityDetail(activityId) {
        return this.post(`${BASE}/leekmms/activity/query/detail`,
            { json: { activity_id: activityId, query_source: 1 }, referer: ACT_REFERER });
    }
    async getEligibleGoods(activityId, opts = {}) {
        const { goodsId = 0, goodsName = '', page = 1, pageSize = 10 } = opts;
        const data = { activity_id: activityId, goods_id: goodsId, goods_name: goodsName,
                       page_number: page, eligible: 1, page_size: pageSize };
        return this.post(`${BASE}/leekmms/activity/query/eligible/goods/listV2`, { json: data, referer: ACT_REFERER });
    }
    async getPriceRules(activityId, goodsIdList) {
        return this.post(`${BASE}/leekmms/activity/query/enroll/rule_v3`,
            { json: { activity_id: activityId, goods_id_list: goodsIdList }, referer: ACT_REFERER });
    }
    async getSuggestPrices(activityId, goodsIdList) {
        const data = { activity_id: activityId, goods_id_list: goodsIdList, source_type: 'PROMO-HomeModule' };
        return this.post(`${BASE}/leekmms/activity/query/enroll/rule_suggest_price`, { json: data, referer: ACT_REFERER });
    }
    async doEnroll(activityId, goodsVolist) {
        const url = `${BASE}/lakemms/enrollV2`;
        const body = { activity_id: activityId, goods_volist: goodsVolist };
        let result = await this.post(url, { json: body, referer: ACT_REFERER });
        if (!result) return null;
        const errorCode = result.error_code;
        if (errorCode === 30001) {
            console.info('[activity] 30001 建议价过期，刷新后重试...');
            const ids = goodsVolist.map(g => g.goods_id).filter(Boolean);
            if (ids.length) await this.getSuggestPrices(activityId, ids);
            result = await this.post(url, { json: body, referer: ACT_REFERER });
        } else if (errorCode === 2002690) {
            console.info('[activity] 2002690 库存超限，去库存后重试...');
            const stripped = goodsVolist.map(g => { const ng = Object.assign({}, g); delete ng.quantity; delete ng.sku_quantity; return ng; });
            result = await this.post(url, { json: { activity_id: activityId, goods_volist: stripped }, referer: ACT_REFERER });
        }
        return result;
    }
}

module.exports = { ActivityEnrollAPI };

File: python/apis/activity_enroll.py
# -*- coding: utf-8 -*-
"""活动报名 API（6 端点，来自 05_activity_enroll.json）"""
import logging
from typing import Optional, Dict, Any, List
from base_request import BaseRequest

BASE = "https://mms.pinduoduo.com"
ACT_REFERER = "https://mms.pinduoduo.com/activity"

logger = logging.getLogger("pdd_activity")


class ActivityEnrollAPI(BaseRequest):
    """活动报名接口"""

    def check_login(self) -> Optional[Dict[str, Any]]:
        """登录状态探针 → POST /janus/api/checkLogin"""
        url = f"{BASE}/janus/api/checkLogin"
        return self.post(url, json_data={}, referer=ACT_REFERER)

    def get_activity_detail(self, activity_id: int) -> Optional[Dict[str, Any]]:
        """查询活动详情 → POST /leekmms/activity/query/detail

        Args:
            activity_id: 活动 ID
        """
        url = f"{BASE}/leekmms/activity/query/detail"
        return self.post(url, json_data={"activity_id": activity_id, "query_source": 1},
                         referer=ACT_REFERER)

    def get_eligible_goods(self, activity_id: int, goods_id: Optional[int] = None,
                           goods_name: str = "", page: int = 1,
                           page_size: int = 10) -> Optional[Dict[str, Any]]:
        """查询可报名商品 → POST /leekmms/activity/query/eligible/goods/listV2"""
        url = f"{BASE}/leekmms/activity/query/eligible/goods/listV2"
        data = {
            "activity_id": activity_id,
            "goods_name": goods_name,
            "page_number": page,
            "eligible": 1,
            "page_size": page_size,
        }
        if goods_id is not None:
            data["goods_id"] = goods_id
        return self.post(url, json_data=data, referer=ACT_REFERER)

    def get_price_rules(self, activity_id: int,
                        goods_id_list: List[int]) -> Optional[Dict[str, Any]]:
        """获取价格规则 → POST /leekmms/activity/query/enroll/rule_v3

        Args:
            activity_id: 活动 ID
            goods_id_list: 商品 ID 列表
        """
        url = f"{BASE}/leekmms/activity/query/enroll/rule_v3"
        return self.post(url,
                         json_data={"activity_id": activity_id, "goods_id_list": goods_id_list},
                         referer=ACT_REFERER)

    def get_suggest_prices(self, activity_id: int,
                           goods_id_list: List[int]) -> Optional[Dict[str, Any]]:
        """获取建议价 → POST /leekmms/activity/query/enroll/rule_suggest_price"""
        url = f"{BASE}/leekmms/activity/query/enroll/rule_suggest_price"
        data = {
            "activity_id": activity_id,
            "goods_id_list": goods_id_list,
            "source_type": "PROMO-HomeModule",
        }
        return self.post(url, json_data=data, referer=ACT_REFERER)

    def do_enroll(self, activity_id: int,
                  goods_volist: List[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
        """执行活动报名 → POST /lakemms/enrollV2

        Args:
            activity_id: 活动 ID
            goods_volist: 报名商品列表，每项含 goods_id, enroll_rule_param, sku_price_list 等

        内置重试策略：
        - error_code 30001（建议价过期）: 自动刷新建议价后重试一次
        - error_code 2002690（库存超限）: 去掉库存参数后重试一次
        """
        url = f"{BASE}/lakemms/enrollV2"
        body = {"activity_id": activity_id, "goods_volist": goods_volist}
        result = self.post(url, json_data=body, referer=ACT_REFERER)

        if not result:
            return None

        error_code = result.get("error_code")
        if error_code == 30001:
            logger.info("error_code=30001（建议价过期），尝试刷新建议价后重试...")
            goods_ids = [g.get("goods_id") for g in goods_volist if g.get("goods_id")]
            if goods_ids:
                self.get_suggest_prices(activity_id, goods_ids)
            return self.post(url, json_data=body, referer=ACT_REFERER)
        if error_code == 2002690:
            logger.info("error_code=2002690（库存超限），去掉库存参数后重试...")
            stripped = []
            for g in goods_volist:
                ng = dict(g)
                ng.pop("quantity", None)
                ng.pop("sku_quantity", None)
                stripped.append(ng)
            return self.post(url, json_data={"activity_id": activity_id,
                                             "goods_volist": stripped}, referer=ACT_REFERER)
        return result
